oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/api/binary/BinaryHttpUpload.java

```diff
@@ -0,0 +1,163 @@
+ *     (integer division, discard modulo which will be the last part)</li>
+ *     <li>upload: segment the binary into partSize, for each segment take the next URL from partURLs
+ *     (strictly in order!), proceed with a standard HTTP PUT for each, and for the last part use whatever
+ *     segment size is left</li>
+ *     <li>if a segment fails during upload, retry (up to a certain time out)</li>
+ *     <li>after the upload has finished successfully, notify the application, for example through a
+ *     complete request, passing the {@link #getUploadToken() upload token}, and the application will have to
+ *     call {@link HttpBinaryProvider#completeHttpUpload(String)} with the token</li>
+ * </ol>
+ *
+ * <h2>JSON view</h2>
+ *
+ * A JSON representation of this interface as passed back to a remote client might look like this:
+ * <pre>
+ * {
+ *     "uploadToken": "aaaa-bbbb-cccc-dddd-eeee-ffff-gggg-hhhh",
+ *     "minPartSize": 10485760,
+ *     "maxPartSize": 104857600,
+ *     "partURLs": [
+ *         "http://server.com/upload/1",
+ *         "http://server.com/upload/2",
+ *         "http://server.com/upload/3",
+ *         "http://server.com/upload/4"
+ *     ]
+ * }
+ * </pre>
+ * </p>
+ *
+ */
+public interface BinaryHttpUpload {
+
+    /**
+     * Returns a token identifying this upload. This is required to finalize the upload
+     * at the end by calling {@link HttpBinaryProvider#completeHttpUpload(String)}.
+     *
+     * <p>
+     * The format of this string is depending on the implementation and opaque to consumers.
+     * Implementations must ensure that clients cannot spoof tokens.
+     *
+     * @return a string token identifying this particular upload
+     */
+    String getUploadToken();
+
+    /**
+     * Return the smallest possible part size in bytes. If a consumer wants to choose a custom
+     * part size, it cannot be smaller than this value.
+     *
+     *
+     * @see #getMaxPartSize() getMaxPartSize() for the maximum
+     *
+     * @return smallest possible part size in bytes
+     */
+    long getMinPartSize();
+
+    /**
+     * Return the largest possible part size in bytes. If a consumer wants to choose a custom
+     * part size, it cannot be larger than this value. If this returns -1, the maximum is
+     * unlimited.
+     *
+     * @see #getMinPartSize() getMinPartSize() for the minimum
+     *
+     * @return largest possible part size in bytes or -1 if unlimited
+     */
+    long getMaxPartSize();
+
+    /**
+     * Returns one or more URLs for uploading the binary in one or more parts.
+     *
+     * <p>
+     * In case of multiple URLs, the order in which they are returned is important and defines
+     * the order of the parts. Each part covers a given range of the binary in bytes.
+     *
+     * @return one or more URLs for uploading the binary in one or more parts
+     */
+    @Nonnull
+    Collection<URL> getUploadURLs();
```

Comment:
We debated use of `String` at one point, because it will just be passed into json or similar for a remote client.

The presign method in S3 used in the implementation returns a URL and for the tests URL was also more practical.

Reply:
You can easily convert a URI to a URL if you need to. That you need a URL somewhere else should not dictate what we expose in our API.

Reply:
What about URL to URI conversion? Does it involve re-parsing?

Reply:
Yes. Is this a problem? (https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#toURI())

Reply:
@reschke I'm not opposed to making the change from URL to URI, but I'm not sure what exactly the benefit is.
If the back-end storage is `S3DataStore`, the library will give us a URL which would have to be converted to a URI.  And then a client wanting to do the upload or download would then have to convert that URI back to a URL in order to use `HttpURLConnection` or `HttpsURLConnection` - or use a library to do it, like Apache HttpComponents.  I don't see the value of converting it to a URI if what a client needs to create a connection is a URL anyway.

Reply:
- Apache HTTPClient wants a URI, not a URL
- URL is a complex object that comes with some legacy baggage, see, for instance,  https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#equals(java.lang.Object)

Reply:
As mentioned, there is no need for production code to consume it directly on the Java side in e.g. Apache HTTPClient, but it will be put as-is into JSON or some other response format to be sent to a remote client.

I know that URL has it's weirdness in some usage scenarios (say building a URL out of pieces which it doesn't support well), but they don't really apply here. For this API, use of URL is pretty opaque, we are just passing through a string URL generated by some libraries to some response format like JSON.

AFAICS, converting from [URL to URI](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/net/URL.java#l939) does [involve](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/net/URI.java#l588) re-parsing the string, which is a small, but unnecessary extra cost.

I would argue a `String` represents this opaqueness best, but I also see the point of @mattvryan saying `URL` as more specific type makes it more clear in the API.

Reply:
That doesn't make any sense. If you put the identifiers into JSON; they will be serialized and re-parsed anyway.

The Java API design should be based on Java's needs. Don't use URL. There's a reason why we don't use it in the JCR API.

Reply:
The URL to URI conversion would be in _addition_. It's not much, but I think it's good to minimize unnecessary string operations & duplications in Java.

Note that for `HttpBinaryProvider.getHttpDownloadURL()` (to which your comment applies as well I assume), this will be a frequently called operation, as it gets invoked for every image on say an asset browsing UI (btw, that's why we have a [cache for the presigned GET URLs](https://github.com/mattvryan/jackrabbit-oak/blob/f46f5802e3dc48e1e3c26e2a5f89cbf3abe0ed8a/oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java#L125) in the DS impls).

Since we start with an URL on the [AWS SDK side](https://github.com/mattvryan/jackrabbit-oak/blob/f46f5802e3dc48e1e3c26e2a5f89cbf3abe0ed8a/oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java#L863), if we have to convert, there should be a good reason. What is the particular issue you see with URL in this case? Issues the AWS SDK might be missing?

"In this case" meaning: in 99.9% of cases, an API client would just do `url.toString()` into JSON. There _might_ be some rare cases where it would inspect the fields, such as the `url.getHost()` or `url.getProtocol()`.

Reply:
If in 99,99% of the cases, the URL gets put into JSON, why does it matter to you? Once again, the performance argument doesn't make *any* sense.

Reply:
Based on the [JDK8 sources](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/net/URL.java): In the S3 case, we start with a `java.net.URL` from the SDK containing all the parts as separate strings. Upon converting to `java.net.URI`, it creates a `StringBuffer` and a `String` with the entire URL as one string in `URL.toString()`. `new URI(String)` will immediately parse it again into the parts. When the client later calls `URI.toString()`, it creates another new `StringBuffer` and `String`. This is an extra parsing round plus an extra copy in `StringBuffer` (not the more efficient `StringBuilder`) and `String` in memory.

In the Azure case, we build the `String` ourselves and can create a `URL` or `URI` as needed, so it doesn't make a big difference there. (Unless we chose to go with `String` in the API).

These URLs are not short, they will be about 300 to 500 chars long. In case of multipart upload in `getUploadURLs()` here, we expect usually 10 to 20 URLs in the list, albeit the frequency is bound by the number of files being uploaded times the number of extra binary renditions that might get created as part of ingestion per uploaded binary. In case of download (`HttpBinaryProvider.getHttpDownloadURL()`), it is only one URL at a time, but it might get called for each UI page rendering, so much more frequent, depending on external caching strategies.

Agree that this isn't talking about major performance & memory usage differences, but it's also not something you can [completely ignore](https://twitter.com/iamdevloper/status/926458505355235328).

@reschke It would really help if you could outline the specific problems you see with `URL` in this case, otherwise why the change with a small, but extra cost?

You mentioned the JCR API: AFAICS, it does not use either `java.net.URL` nor `URI` anywhere, instead it uses plain `String` for namespace prefixes and URI type values. Using this precedence, could we agree on switching to plain `String` here?